### DISCUSSION TITLE
Database Transaction: for multiple tables actions in one transaction and rollback if any error occurred.

### DISCUSSION THREAD

**Discussion by JasonChiu-dev** [supabase/supabase#3732] [Questions] [answered]
Hello everyone!

I need the transaction handling from supabase.io tech team as follow:

Regarding BEGIN / COMMIT / ROLLBACK on Supabase,  like:
await supabase.transaction(
supabase.from(...).insert(table1...),
if any error, rollback.
supabase.from(...).update(table2...)
if any error, rollback.
supabase.from(...).updata(table3...)
if no error, Commit.
)

But I got answer from supabase.io tech team that no method to handle transaction. only rpc.

But I don't know how to really implement the correct rpc function for multiple database actions with complex arguments for database actions, including insert to the first table by multiple rows and then update the second table for different rows with different id and then update the third table for updating the data by key, if there is any error occurred during the THREE TABLEs transaction then ROLLBACK, otherwise COMMIT the transaction.

Is there Any solution or suggestion for writing the correct rpc function to handle the multiple tables' operations(insert, update, ...) in one transaction which can make sure the data consistency?

Here is my Flutter sample code for multiple database actions:
`
void addProduct(...) {
  // Todo: Begin transaction here

  // insert data into the first table: table_name1
  final response = await supabase.from('table_name1').insert([
        {'id': '1001', 'name': 'Apple', ‘price’: 5},
        {'id': '1002', 'name': 'Orange', ‘price’: 3},
        {'id': '1003', 'name': 'Pine Apple', ‘price’: 15},
        {'id': '1004', 'name': 'Water Melon', ‘price’: 10},
        {'id': '1005', 'name': 'Sweet Fruit', ‘price’: 5}
      ]).execute();

  if (response.error != null) {
    // Error occurred
    // Todo: Rollback transaction
  } else {
    // update data to the second table: table_name2 corresponding to
    // the products in the first table: table_name_1

    for (int i = 0; i < items.length; i++) {
      double newQuantity = table_name2.items[i].quantity + 
        table_name1.products[i].quantity;

      // update the second table: table_name2
      final response = await supabase
            .from(table_name2)
            .update({'quantity': newQuantity})
            .eq('id', items[i].id)
            .execute();

      if (response.error != null) {
        // Error occurred
        // Todo: break the for loop and Rollback transaction
      } else {
        // update the third table: table_name3 using the data 
        // amountSum  from outside
        double newAmount = table_name3.amount + amountSum; 

        // update the third table: table_name3

        final response = await supabase
            .from('table_name3'
            .update({'amount': newAmount})
            .eq('id', table_name3.id)
            .execute();

        if (response.error != null) {
          // Error catch.
          // Todo: Rollback transaction
        } else {
          // the database actions are all succeed
          // Todo: Commit transaction
        }
      }
    }
  }
}
`

I need your help.

Jason

**Comment by steve-chavez** [comment, marked as answer]
Hi @JasonChiu-dev,

Every call to a postgres function with supabase rpc is wrapped in a transaction. All you need to do to  ROLLBACK is to [RAISE](https://www.postgresqltutorial.com/plpgsql-errors-messages/) an error. Here's an example:
​
```sql
create or replace function three_table_logic() returns void as $$
begin
 -- do some logic here
 -- insert one
 -- insert two
 -- insert three
 if not condition then
  RAISE exception 'an error '
    using detail = 'wrong three table logic',
          hint = 'repeat the operation';
   -- the above will cause a rollback
 end if;
end $$ language plpgsql;
```

If the function above failed, it will show an error on the client:

```js
const { data, error } = await supabase.rpc('three_table_logic')
console.log(error);
{
  "message":"an error",
  "details":"wrong three table logic",
  "hint":"repeat the operation",
  "code":"P0001"
}
```
